apps/examples/src/misc/end-to-end.tsx

```diff
@@ -1,11 +1,50 @@
 import { getLicenseKey } from '@tldraw/dotcom-shared'
-import { useEffect } from 'react'
-import { Tldraw, createShapeId, exportAs, useActions, useEditor } from 'tldraw'
+import { useEffect, useLayoutEffect } from 'react'
+import {
+	BaseBoxShapeUtil,
+	TLBaseShape,
+	Tldraw,
+	createShapeId,
+	exportAs,
+	useActions,
+	useEditor,
+	useSafeId,
+} from 'tldraw'
 import 'tldraw/tldraw.css'
 import { EndToEndApi } from './EndToEndApi'
 ;(window as any).__tldraw_ui_event = { id: 'NOTHING_YET' }
 ;(window as any).__tldraw_editor_events = []
 
+type HtmlCssShape = TLBaseShape<'html', { html: string; css: string; w: number; h: number }>
+class HtmlCssShapeUtil extends BaseBoxShapeUtil<HtmlCssShape> {
+	static override type = 'html'
+
+	override getDefaultProps(): { html: string; css: string; w: number; h: number } {
+		return { w: 100, h: 100, html: '', css: '' }
+	}
+	override component(shape: HtmlCssShape) {
+		return <HtmlCssShapeComponent shape={shape} />
+	}
+	override indicator(shape: HtmlCssShape) {
+		return <rect width={shape.props.w} height={shape.props.h} />
+	}
+}
+function HtmlCssShapeComponent({ shape }: { shape: HtmlCssShape }) {
+	// eslint-disable-next-line local/no-at-internal
+	const id = useSafeId()
+
+	useLayoutEffect(() => {
+		const style = document.createElement('style')
+		style.textContent = shape.props.css.replace(/#self/g, `#${id}`)
```

Comment:
what is this `#self` doing?

Reply:
added a comment for this - it's so test cases can write CSS targeting an individual shape